### PR TITLE
feat(artifacts): Support artifacts filtering in Travis triggers

### PR DESCRIPTION
**Pull request by jervi**
Corresponds with https://github.com/spinnaker/igor/pull/223

### CODE REVIEW

**Comment by jervi** [comment]
@spinnaker/reviewers PTAL

File: echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Artifact extends com.netflix.spinnaker.kork.artifacts.model.Artifact {
+
+  @Override
+  public Map<String, Object> getMetadata() {
```
Comment (jtk54) [inline echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java]: Are these two methods necessary if they are strictly calling the `super` method?

File: echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  @Override
+  public void setMetadata(Map<String, Object> metadata) {
+    super.setMetadata(metadata);
+  }
+
+  // Add extra data in the JSON from Igor to metadata:
+  @JsonAnySetter
+  public void putMetadata(String key, Object value) {
+    if (super.getMetadata() == null) {
```
Comment (jtk54) [inline echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java]: Maybe this logic belongs in the superclass in Kork?

File: echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package com.netflix.spinnaker.echo.model;
+
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Artifact extends com.netflix.spinnaker.kork.artifacts.model.Artifact {
```
Comment (lwander) [inline echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java]: This probably shouldn't live in echo

File: echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  public Map<String, Object> getMetadata() {
+    return super.getMetadata();
+  }
+
+  @Override
+  public void setMetadata(Map<String, Object> metadata) {
+    super.setMetadata(metadata);
+  }
+
+  // Add extra data in the JSON from Igor to metadata:
```
Comment (lwander) [inline echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java]: This feels like it'll surprise us/bite us when strange objects are serialized entirely into artifacts by virtue of being placed into metadata.

**Comment by jervi** [comment]
Thanks Lars and Jacob.
I extended the Kork `Artifact` as I thought of this as a special case not relevant in other places (and it only adds stuff for Jackson deserialization, it doesn't actually change the model), but I have no problem adding it directly to the superclass in Kork instead.
The rationale for the change is that the Jenkins/Travis artifacts from Igor have fields that Kork `Artifact` don't have: https://github.com/spinnaker/igor/blob/master/igor-web/src/main/groovy/com/netflix/spinnaker/igor/build/model/GenericArtifact.groovy
So it seemed like a simple way of supporting arbitrary fields.
A Jenkins/Travis artifact then looks like this in Echo:
```
{
    "metadata": {
        "displayPath": "<somePath>",
        "fileName": "<someFileName>",
        "relativePath": "<someRelativePath>"
    },
    "name": "<name>",
    "reference": "<reference>",
    "type": "<type>",
    "version": "<version>"
}
```

Another solution could be to add `displayPath`, `fileName` and `relativePath` to Kork's `Artifact`?

File: echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java (hunk 1 above)
Comment (jervi) [reply echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/Artifact.java]: True, fixed

**Comment by jervi** [comment]
@jtk54 and @lwander, any preferences on what to do with the missing fields (`displayPath`, `fileName` and `relativePath`)? 

Suggestions:

- Ignore them
- Add the fields to the Kork `Artifact` class
- Put the data in the metadata map
  - As implemented in this PR
  - Or add the `@JsonAnySetter` to the Kork `Artifact` class

**Comment by jtk54** [comment]
@jervi Sorry for the response delay -- I think adding to the metadata as implemented here works well.

**Comment by jervi** [comment]
Ref Slack chat: I've moved the `@JsonAnySetter` to Kork: https://github.com/spinnaker/kork/pull/138

**Comment by jtk54** [comment]
Tested with the new Spinnaker dependenies bump? Good to merge?

**Comment by jervi** [comment]
Yes, it's good to merge 👍

**Pull request merged**